### PR TITLE
Upgrade protobuf to 27.1 and modernize codegen using new APIs (e.g. `has_presence()`)

### PR DESCRIPTION
**Pull request by benjaminp**
Fixes https://github.com/grpc/grpc-web/issues/1437.

### CODE REVIEW

**Comment by sampajano** [comment]
@benjaminp Wow thanks so much for help working on this!

I just ran our CI and ran into this error tho:

https://btx.cloud.google.com/invocations/d6c6aeb6-b46f-4bca-a17a-6be28b1edbb2/targets/grpc%2Fweb%2Fpresubmit/log

**Comment by benjaminp** [comment]
I've noticed that older GCC versions give a false deprecation warning in part of protobuf, which then explodes because protobuf passes `-Werror`. I can throw `-Wno-deprecated-declarations` in `.bazelrc`.

**Comment by sampajano** [comment]
> I've noticed that older GCC versions give a false deprecation warning in part of protobuf, which then explodes because protobuf passes `-Werror`. I can throw `-Wno-deprecated-declarations` in `.bazelrc`.

oh wow! good to note!! Thanks for the suggestion!

Do you think `-Wno-error=deprecated-declarations` might work equally (but doesn't completely remove warning messages) too?

**Comment by sampajano** [comment]
> I've noticed that older GCC versions give a false deprecation warning in part of protobuf, which then explodes because protobuf passes `-Werror`. I can throw `-Wno-deprecated-declarations` in `.bazelrc`.

Hmm CI still showing same errors it seems:

https://btx.cloud.google.com/invocations/e1797645-dda6-4134-885a-6daf896213a7/targets/grpc%2Fweb%2Fpresubmit/log

**Comment by benjaminp** [comment]
> Do you think `-Wno-error=deprecated-declarations` might work too (but doesn't completely remove warning messages)?

Should work, yes.

> Hmm CI still showing same errors it seems:
> 
> https://btx.cloud.google.com/invocations/e1797645-dda6-4134-885a-6daf896213a7/targets/grpc%2Fweb%2Fpresubmit/log

Okay, maybe I need to also pass `--host_copt` as in the latest revision?

**Comment by sampajano** [comment]
Thanks for the update!!

Still same error tho... the bazel flag might not have been working..

Does it make things build for you?

When running this local command (on my Mac) shows me the same error as in CI:
`docker-compose up --build prereqs`

**Comment by sampajano** [comment]
Aha got it!!

You need to add an extra line:

```
COPY ./.bazelrc ./.bazelrc
```

below this line:

https://github.com/grpc/grpc-web/blob/8ab32b945ca0530222593127e9431a455eff24a8/net/grpc/gateway/docker/prereqs/Dockerfile#L77

Then it should work :)

**Comment by benjaminp** [comment]
Nice find. done

File: javascript/net/grpc/web/generator/grpc_generator.cc [hunk 1]
```diff
@@ -841,13 +841,11 @@ void PrintProtoDtsMessage(Printer* printer, const Descriptor* desc,
                      "set$js_field_name$(value?: $js_field_type$): "
                      "$class_name$;\n");
     }
-    if (field->has_optional_keyword() ||
-        (field->type() == FieldDescriptor::TYPE_MESSAGE &&
-            !field->is_repeated() && !field->is_map())) {
+    if (field->has_presence()) {
```
Comment (sampajano) [inline javascript/net/grpc/web/generator/grpc_generator.cc:844]: what a simplification! 😊

File: javascript/net/grpc/web/generator/grpc_generator.cc [hunk 2]
```diff
@@ -867,14 +865,12 @@ void PrintProtoDtsMessage(Printer* printer, const Descriptor* desc,
...
-  for (int i = 0; i < desc->oneof_decl_count(); i++) {
-    const OneofDescriptor* oneof = desc->oneof_decl(i);
-    if (!oneof->is_synthetic()) {
-      vars["js_oneof_name"] = ToUpperCamel(ParseLowerUnderscore(oneof->name()));
-      printer->Print(
-          vars, "get$js_oneof_name$Case(): $class_name$.$js_oneof_name$Case;\n");
-      printer->Print("\n");
-    }
+  for (int i = 0; i < desc->real_oneof_decl_count(); i++) {
+    const OneofDescriptor *oneof = desc->real_oneof_decl(i);
```
Comment (sampajano) [inline javascript/net/grpc/web/generator/grpc_generator.cc:869]: nice! 😊

File: javascript/net/grpc/web/generator/grpc_generator.cc [hunk 3]
```diff
@@ -904,8 +900,7 @@ void PrintProtoDtsMessage(Printer* printer, const Descriptor* desc,
     }
     vars["js_field_name"] = js_field_name;
     vars["js_field_type"] = AsObjectFieldType(field, file);
-    if ((field->type() != FieldDescriptor::TYPE_MESSAGE && !field->has_optional_keyword()) ||
```
Comment (sampajano) [inline javascript/net/grpc/web/generator/grpc_generator.cc:907]: Trying to understand all the cases is literally giving me a headache.. 😅

so I asked ChatGPT if the before and after is equivalent and it says yet — so i trust it (and you) :D

I honestly wouldn't be surprised if the code before has bugs which would be fixed by the new code.

thanks so much for the simplification.. so much nicer :)

**Review by sampajano — APPROVED** [review]
Thank you SO MUCH for this contribution.. This is wonderful.. such a great simplification across the board and I much appreciate you help us make our lives easier using these new APIs.. Thank you 😊

---

Only one request — to remove the `--host_copt` — other comments are all just thoughts :)

Thank you again :)

File: .bazelrc [hunk 4]
```diff
@@ -0,0 +1 @@
+build --copt=-Wno-error=deprecated-declarations --host_copt=-Wno-error=deprecated-declarations
```
[the remark this answers is not in the document]
Comment (sampajano) [reply .bazelrc:1]: @benjaminp Could you update this file? I'll re-run CI and merge this PR afterwards.. Thanks! 😊

[the remark this answers is not in the document]
Comment (benjaminp) [reply .bazelrc:1]: I think it's probably safest to keep this, and there's minimal downside. If any protobuf code ends up in the Bazel exec configuration, the compilation problem will reappear.

[the remark this answers is not in the document]
Comment (sampajano) [reply .bazelrc:1]: Ok! Sure! I think you understand bazel (and protobuf for the matter) way more than me so I trust your judgement here :)

**Comment by sampajano** [comment]
@benjaminp Thank you SO much for the wonderful contribution here again :)

Let me know if you need a new release to be cut to be able to leverage some of the change here, and if so i'd be happy to do one soon (after July 4th tho) :)

**Comment by sampajano** [comment]
FYI i think this has broken:

`docker-compose up --build protoc-plugin`

Error logs (internal):

```
 /bin/mkdir -p '/usr/local/bin'
  /bin/bash ../libtool   --mode=install /usr/bin/install -c protoc '/usr/local/bin'
libtool: install: /usr/bin/install -c .libs/protoc /usr/local/bin/protoc
make[2]: Leaving directory '/github/grpc-web/third_party/protobuf/src'
make[1]: Leaving directory '/github/grpc-web/third_party/protobuf/src'
Removing intermediate container f2b2b179b086
 ---> d990ba2c668c
Step 6/6 : RUN cd ./javascript/net/grpc/web/generator &&   make protoc-gen-grpc-web
 ---> Running in 52e698a00ba2
g++ -std=c++11 -I/usr/local/include -pthread  -c -o grpc_generator.o grpc_generator.cc
grpc_generator.cc: In function 'void grpc::web::{anonymous}::PrintProtoDtsMessage(google::protobuf::io::Printer*, const google::protobuf::Descriptor*, const google::protobuf::FileDescriptor*)':
grpc_generator.cc:869:42: error: 'const class google::protobuf::Descriptor' has no member named 'real_oneof_decl'; did you mean 'oneof_decl'?
  869 |     const OneofDescriptor *oneof = desc->real_oneof_decl(i);
      |                                          ^~~~~~~~~~~~~~~
      |                                          oneof_decl
make: *** [<builtin>: grpc_generator.o] Error 1
Service 'protoc-plugin' failed to build: The command '/bin/sh -c cd ./javascript/net/grpc/web/generator &&   make protoc-gen-grpc-web' returned a non-zero code: 2
```

**Comment by sampajano** [comment]
I've temporarily `protoc-plugin` in CI in https://github.com/grpc/grpc-web/pull/1447.

Would like to have that fixed soon tho!

@benjaminp If you could help take a look it'll be appreciated too!

**Pull request merged**